pytest/tests/mocknet/mirror.py

```diff
@@ -576,25 +607,14 @@ def run_env_cmd(args, traffic_generator, nodes):
         'update-binaries',
         help=
         'Update the neard binaries by re-downloading them. The same urls are used.'
-    )
+        + ' Add or override the URLs with --neard-binary-url.')
+    update_binaries_parser.add_argument('--neard-binary-url', type=str)
+    update_binaries_parser.add_argument('--update-epoch-height', type=int)
```

Comment:
I think `epoch_height` is probably not the right parameter for what we want to do here. So if you give a `neard_upgrade_binary_url` when running `init-neard-runner`, then all nodes will have two binaries, `neard0` and `neard1`, and the epoch height parameter you're referencing above in `neard_runner.py` is set to `0` for `neard0` (meaning start with that binary from the beginning), and then each node gets a random epoch height at which to upgrade to `neard1`. So if you set `epoch_height=1`, then you're modifying `neard1` for the nodes who happened to be configured to upgrade to `neard1` at `epoch_height=1`, but for all the other nodes we're adding another binary to be upgraded to at `epoch_height=1`, leaving us with three binaries `neard0` `neard1` and `neard2`. And based on the logic added in `neard_runner.py` here, actually `neard2` will have the epoch height at which we want to upgrade to it set to `1`, but `neard1` will have an epoch height higher than `1`, which means we will upgrade to `neard2` (at, say, epoch height `3`), but then immediately upgrade to `neard1` since at that point we're past epoch height `1`.

What you probably want to refer to in this flag is the index of the binary we're updating. We want to say here whether we want to upgrade `neard0` or `neard1`. The epoch height is just a piece of info used to tell when to actually upgrade it

Reply:
The intention is to pass the epoch height where you want to upgrade to the new binary. We want as well to amend the binary for a certain epoch height if the epoch height exists.

Reply:
```
actually neard2 will have the epoch height at which we want to upgrade to it set to 1, but neard1 will have an epoch height higher than 1
```
Not sure that I understand what you mean here. When new binaries are downloaded, neard_runner will first sort the binaries by the epoch height, so give epoch heights: [0,1 3] we will end up with 3 binaries: `neard0` for epoch 0, `neard1` for epoch 1, `neard2` for epoch 3

Reply:
Oh yea I forgot about that sorting. But still we end up with different numbers of binaries on different hosts. If you start a forknet with 2 nodes, and ask for an upgrade binary, then let's say the binaries config on `node0` looks like `[{"url": <url0>, "epoch_height": 0}, {"url": <url1>, "epoch_height": 2}]` and on `node1` it looks like: `[{"url": <url0>, "epoch_height": 0}, {"url": <url1>, "epoch_height": 3}]`.

Then you run `mirror.py update-binaries --neard-binary-url <url2> --update-epoch-height 2`

Then on `node0` the config is now `[{"url": <url0>, "epoch_height": 0}, {"url": <url2>, "epoch_height": 2}]`, and on `node1` the config is now `[{"url": <url0>, "epoch_height": 0}, {"url": <url2>, "epoch_height": 2}, {"url": <url1>, "epoch_height": 3}]`

This is surely not what we want right? two binaries on one of the nodes and three on the other. What instead makes sense to me is to tell me which index in the sequence of binaries to upgrade to (that is, do you want to upgrade `neard0` or `neard1`). If you want to also include an option to set the epoch height, I think that's fine, but keep in mind that that was always meant to be different on different nodes, so we can test what happens when different versions are running on the network

Reply:
I see your point of having this option to update a certain binary by idx.
This command was intended for stakewars scenario where we want to update the binaries at a certain height in the future and not go back to a backup.

Reply:
I don't see a clean way to add a cmd to update the binary at a certain index. If we go ahead with epoch height pram, we could end up in the scenario described by you and then the idx param will be even more confusing. a sad scenario would be: we use height to update binaries and ended up with some hosts having [ver 1, ver 3, ver 2] and on some others having [ver 1, ver 2, ver 3]. In this case, using `idx` will not make much sense either.

Reply:
@marcelo-gonzalez, I will add some comment to explain the behaviour of inserting the epoch height.

Reply:
I think there is some confusion here. What I'm describing is just just a difference in style, or an extra feature I want, but actually just the most natural way to implement it. The scenario I described above is just undesirable and unexpected.

For a second let's forget about the epoch height parameter, supposing that the user does not care to change that. And let's again go with the example I posted above, with two binaries, where `node0` has binaries config `[{"url": <url0>, "epoch_height": 0}, {"url": <url1>, "epoch_height": 2}]` and `node1` has binaries config `[{"url": <url0>, "epoch_height": 0}, {"url": <url1>, "epoch_height": 3}]`

Then if instead of the `--epoch-height` flag, we have a `--binary-index` flag (or some other name as you see fit), then when we run `mirror.py update-binaries --neard-binary-url <url2> --binary-index 1`, we will have this on `node0`: `[{"url": <url0>, "epoch_height": 0}, {"url": <url2>, "epoch_height": 2}]` and this on `node1`: `[{"url": <url0>, "epoch_height": 0}, {"url": <url2>, "epoch_height": 3}]`

Reply:
I also think that it may not be trivial to identify the right epoch height, given that you need to know the start height and the fixed epoch length, and the epoch length may also change right? What if we remove this extra parameter and allow the command to update the binary immediately? and if the network is still running fail to execute the command (the person invoking the command should make sure the nodes are not running neard at the time of running the command).

Reply:
@tayfunelmas, the epoch length is not a concern here as nead-runner will check the current epoch height(epoch number) and load the right binary.

Reply:
@marcelo-gonzalez, the user can exercise caution when running the command with epoch height to not end up with inconsistent binary list in config. 

I do believe that this is a useful feature if you want to test a protocol upgrade without resetting the network.

Reply:
@marcelo-gonzalez, the scenario that you described is not the target of this PR. Indeed, if you want to upgrade the upgrade_binary_url (the seconds neard binary), you need to use the index no the epoch height, hence, a different feature.

Reply:
Sure sounds good, I think we can do the index thing in a separate PR I guess

> the user can exercise caution when running the command with epoch height to not end up with inconsistent binary list in config.

I think adding a flag that will almost always result in an incorrect/undesirable state and then reasoning that the user should just exercise caution when using it is not the way to go. The commands should be easy to understand and just work and do the thing the user intends. We should be leaning towards easy to use/hard to misuse APIs that are very clear, not towards APIs that are almost always going to be incorrect unless certain conditions are met.

I just do not understand why the behavior of `--upgrade-epoch-height` in this PR is desirable. What if we just remove it? And for now we just always upgrade the URL of `neard0`, and don't touch the epoch height. I think for our current purposes that would be sufficient, since that's what we've been doing manually anyway.

@VanBarbascu  I guess, could you just explain the ideal flow you had in mind with this `--upgrade-epoch-height` flag? Let's say we're running a network with 20 nodes, and there are are two binaries on all the hosts, `neard0` and `neard1`. And let's say 5 nodes set epoch height of `neard1` to `1`, another 5 set it to `2`, another 5 set it to `3` and another 5 set it to `4`. What did you have in mind as to what we are trying to achieve when we run  `mirror.py update-binaries --neard-binary-url <url2> --update-epoch-height {some_height}` ?

Reply:
I don't understand why is there such a high friction for this feature that can go in parallel with the index based update. 

 I may repeat myself, --upgrade-epoch-height argument is not for the case where you have staggered epoch heights but for the case where you want to do a protocol upgrade without resetting the network to epoch 0.
I.E. you are at epoch 5 and you want to upgrade binaries at epoch 7. You can insert a new binary for epoch 7. 

I am saying that the two args `--upgrade-epoch-height` and `--upgrade-binary-idx` can be both used because we cannot add future binary upgrades only with binary idx arg.

For this case that you keep presenting, the right solution is indeed to use the binary index argument.

Reply:
> I don't understand why is there such a high friction for this feature that can go in parallel with the index based update.

Because the current PR introduces behavior that is not usually desired and will be unexpected for most people. I don't actually care that much about the `--upgrade-binary-idx` feature. We can do it in the future if we want but it's not a huge deal. What I care about is introducing flags that will do unexpected things and be frustrating to people because they lead to a strange state

> For this case that you keep presenting, the right solution is indeed to use the binary index argument.

The thing is, this situation is a common one. It's actually the standard one. When you say "--upgrade-epoch-height argument is not for the case where you have staggered epoch heights", I think what you're missing is that we *do* have staggered epoch heights whenever we run this test for releases and have two binaries (neard0 and neard1). We can't just add a flag and hope that people will only use it if they've correctly determined that the flag is safe for their case

> I may repeat myself, --upgrade-epoch-height argument is not for the case where you have staggered epoch heights but for the case where you want to do a protocol upgrade without resetting the network to epoch 0.
I.E. you are at epoch 5 and you want to upgrade binaries at epoch 7. You can insert a new binary for epoch 7.

Okay, I think this is where the confusion came from. Nowhere in this PR's description or comments elsewhere was this behavior described. From my perspective as a reviewer I have to just try and guess at what was intended, and it looked incorrect to me.

And I think this behavior was unexpected to me because as far as I remember, we have never wanted to do this. Have we wanted to do this recently? What we've been doing recently is just upgrading a single binary (`neard0`). Imagine someone trying to use this tool who has not taken a look at this PR. Do you think it would be obvious how they should construct the command to do the very simple thing of upgrading `neard0`? You'd have to go: `mirror.py update-binaries --neard-binary-url <url> --update-epoch-height 0`. How would you even know what this epoch height thing is?

In any case, if this is a useful feature that we want, I still think the UX needs to be improved so people don't shoot themselves in the foot with this flag. Can we please change things so that this is more clear by adding an extra subcommand instead? here is a diff:

```diff
diff --git a/pytest/tests/mocknet/mirror.py b/pytest/tests/mocknet/mirror.py
index d575bd2e3..e8f1a2352 100755
--- a/pytest/tests/mocknet/mirror.py
+++ b/pytest/tests/mocknet/mirror.py
@@ -429,10 +429,12 @@ def start_traffic_cmd(args, traffic_generator, nodes):
 
 
 def update_binaries_cmd(args, traffic_generator, nodes):
-    if (args.neard_binary_url or args.update_epoch_height
-       ) and not (args.neard_binary_url and args.update_epoch_height):
-        logger.warning(f'specify both neard_binary_url and update_epoch_height')
-        return
+    pmap(
+        lambda node: node.neard_runner_update_binaries(
+            args.neard_binary_url, 0), nodes + to_list(traffic_generator))
+
+
+def add_binary_cmd(args, traffic_generator, nodes):
     pmap(
         lambda node: node.neard_runner_update_binaries(
             args.neard_binary_url, args.update_epoch_height),
@@ -614,24 +616,37 @@ if __name__ == '__main__':
                                                    help='''
         Update the neard binaries by re-downloading them. The same urls are used.
         Add or override the URLs with --neard-binary-url by specifying the epoch height.
-        If the network was started with 2 urls, the upgrade epoch height can be randomly assigned
-        on each host. Use caution when specifying --update-epoch-height so that it will not interlace
-        with the random upgrade window for neard1.
         ''')
     update_binaries_parser.add_argument('--neard-binary-url',
                                         type=str,
                                         help='''
+        The URL from which to download the default neard binary (neard0).
         If you plan to restart the network multiple times, it is recommended to use
         URLs that only depend on the branch name. This way, every time you build,
         you will not need to update the URL but just run update-binaries.
         ''')
-    update_binaries_parser.add_argument('--update-epoch-height',
-                                        type=int,
-                                        help='''
+    update_binaries_parser.set_defaults(func=update_binaries_cmd)
+
+    add_binary_parser = subparsers.add_parser('add-binary',
+                                              help='''
+        Add an extra binary to the test nodes, to be upgraded to at the given epoch height
+        ''')
+    add_binary_parser.add_argument('--neard-binary-url',
+                                   type=str,
+                                   required=True,
+                                   help='''
+        If you plan to restart the network multiple times, it is recommended to use
+        URLs that only depend on the branch name. This way, every time you build,
+        you will not need to update the URL but just run update-binaries.
+        ''')
+    add_binary_parser.add_argument('--update-epoch-height',
+                                   type=int,
+                                   required=True,
+                                   help='''
         The epoch height where this binary will begin to run.
         If a binary already exists on the host for this epoch height, the old one will be replaced.
         ''')
-    update_binaries_parser.set_defaults(func=update_binaries_cmd)
+    add_binary_parser.set_defaults(func=add_binary_cmd)
 
     run_cmd_parser = subparsers.add_parser('run-cmd',
                                            help='''Run the cmd on the hosts.''')

```

Reply:
Great suggestion! I've added a new command called `amend-binaries` that has both options: index based and epoch height based update.